external/js/cothority/src/v2/byzcoin/transactionBuilder.ts

```diff
@@ -0,0 +1,143 @@
+        this.instructions = [];
+        return [ctx, response];
+    }
+
+    /**
+     * @return true if one or more instructions are available.
+     */
+    hasInstructions(): boolean {
+        return this.instructions.length > 0;
+    }
+
+    /**
+     * Appends a new instruction.
+     *
+     * @param inst the new instruction to append
+     * @return the appended instruction
+     */
+    append(inst: Instruction): Instruction {
+        this.instructions.push(inst);
+        return inst;
+    }
+
+    /**
+     * Prepends a new instruction
+     *
+     * @param inst the instruction to prepend
+     * @return the prepended instruction
+     */
+    prepend(inst: Instruction): Instruction {
+        this.instructions.unshift(inst);
+        return inst;
+    }
+
+    /**
+     * Appends a spawn instruction.
+     *
+     * @param iid the instance ID where the instruction is sent to
+     * @param contractID the contractID to spawn
+     * @param args arguments of the contract
+     * @return new instruction - can be used for deriveID
+     */
+    spawn(iid: Buffer, contractID: string, args: Argument[]): Instruction {
+        return this.append(Instruction.createSpawn(iid, contractID, args));
+    }
+
+    /**
+     * Appends an invoke instruction.
+     *
+     * @param iid the instance ID where the instruction is sent to
+     * @param contractID the contractID to invoke
+     * @param command to be invoked on the contract
+     * @param args arguments of the command
+     * @return new instruction - can be used for deriveID
+     */
+    invoke(iid: Buffer, contractID: string, command: string, args: Argument[]): Instruction {
+        return this.append(Instruction.createInvoke(iid, contractID, command, args));
+    }
+
+    /**
+     * Appends a delete instruction.
+     *
+     * @param iid the instance ID where the instruction is sent to
+     * @param contractID to be deleted - must match the actual contract
+     * @return new instruction - can be used for deriveID
+     */
+    delete(iid: Buffer, contractID: string): Instruction {
+        return this.append(Instruction.createDelete(iid, contractID));
+    }
+
+    /**
+     * Returns the ID that will be produced by the given instruction
+     *
+     * @param index of the instruction
+     * @param name if given, will be passed to deriveID
+     */
+    deriveID(index: number, name = ""): InstanceID {
+        if (index < 0 || index > this.instructions.length) {
+            throw new Error("instruction out of bound");
+        }
+        return this.instructions[index].deriveId(name);
+    }
+
+    /**
+     * returns a useful and readable representation of all instructions in this transaction.
+     */
+    toString(): string {
+        return this.instructions.map((inst, i) => {
+            const t = ["Spawn", "Invoke", "Delete"][inst.type];
+            let cid: string;
+            let args: Argument[];
+            switch (inst.type) {
```

Comment:
Consider updating return type of `Instruction.get type` & remove `Instruction.type*` to smth like `type InstructionType = "spawn" | "invoke" | "delete"`

Reply:
> but that doesn't seem to be correct TS

Yep, as `Instruction.typeSpawn` is a value of type `number`, not the type `1` (it's kinda close but there is two different DSL, one for values as in JS and one for types for TS).
In the `0 | 1 | 2` case, it creates a type which matches only the `number` value 0, 1 or 2.
In the `"spawn" | "invoke" | "delete"`, it creates a type which matches only the `string` value "spawn", "invoke" or "delete".

> So I kept the numbers - does it look better?

Having the numbers is safer, and using discrete `string` would make it more readable. Having the serialization (ie the representation as `number`) separated from the coder-readable values does allow for more safety and less code-context to keep around. 
Anyway, it's kinda nitpicking, feel free to disregard.